sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -2932,6 +2932,18 @@ class AstBuilder extends DataTypeAstBuilder
     val branches = ctx.whenClause.asScala.map { wCtx =>
       (expression(wCtx.condition), expression(wCtx.result))
     }
+    branches.foreach(
+      branch =>
+        branch._1 match {
+          case _: LambdaFunction =>
+            throw new ParseException(
+              "INVALID_LAMBDA_USAGE",
+              Map("lambdaExpr" -> branch._1.sql),
+              ctx
+            )
+          case _ =>
+        }
+    )
```

Comment:
I don't think this is the right place to do this.
This is not the narrow waist.

Reply:
After a discussion with my team, we concluded that it would be better to fix it as you suggested (to disallow all the options instead of doing it one by one...). Someone will pick it up and finish.